File: RailgunVB/Core/Preconditions/RoleLock.vb
Imports System.Text
Imports Discord
Imports Discord.Commands
Imports Microsoft.Extensions.DependencyInjection
Imports RailgunVB.Core.Enums
Imports TreeDiagram
Imports TreeDiagram.Models.Server

Namespace Core.Preconditions
    
    <AttributeUsage(AttributeTargets.Class Or AttributeTargets.Method)>
    Public NotInheritable Class RoleLock
        Inherits PreconditionAttribute
        
        Private ReadOnly _moduleType As ModuleType
        
        Public Sub New(moduleType As ModuleType)
            _moduleType = moduleType
        End Sub

        Public Overrides Async Function CheckPermissionsAsync(context As ICommandContext, command As CommandInfo, 
                                                        services As IServiceProvider) As Task(Of PreconditionResult)
            Dim user As IGuildUser = context.User
            Dim output As New StringBuilder
            
            Using db As TreeDiagramContext = services.GetService(Of TreeDiagramContext)
                Select _moduleType
                    Case ModuleType.Music
                        Dim data As ServerMusic = Await db.ServerMusics.GetAsync(context.Guild.Id)
                        
                        If data Is Nothing OrElse data.AllowedRoles.Count < 1
                            Return Await Task.FromResult(PreconditionResult.FromSuccess())
                        End If
                        
                        Dim tempOutput As New StringBuilder
                        
                        For Each allowedRole As AllowedRole In data.AllowedRoles
                            Dim role As IRole = context.Guild.GetRole(allowedRole.RoleId)
                            
                            tempOutput.AppendLine($"| {role.Name} |")
                            
                            If user.RoleIds.Contains(allowedRole.RoleId)
                                Return Await Task.FromResult(PreconditionResult.FromSuccess())
                            End If
                        Next
                        
                        output.AppendLine("This command is locked to specific role(s). You must have the following role(s)...") _ 
                            .AppendLine() _
                            .AppendLine(tempOutput.ToString())
                        
                        Exit Select
                End Select
            End Using
            
            Return Await Task.FromResult(PreconditionResult.FromError(output.ToString()))
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Core/Preconditions/RoleLock.cs
using System;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Discord;
using Discord.Commands;
using Microsoft.Extensions.DependencyInjection;
using Railgun.Core.Enums;
using TreeDiagram;

namespace Railgun.Core.Preconditions
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
    public class RoleLock : PreconditionAttribute
    {
        private readonly ModuleType _moduleType;

        public RoleLock(ModuleType modType) => _moduleType = modType;

        public override async Task<PreconditionResult> CheckPermissionsAsync(ICommandContext context, CommandInfo command, IServiceProvider services) {
            var user = (IGuildUser)context.User;
            var output = new StringBuilder();

            using (var db = services.GetService<TreeDiagramContext>()) {
                switch (_moduleType) {
                    case ModuleType.Music:
                        var data = await db.ServerMusics.GetAsync(context.Guild.Id);

                        if (data == null || data.AllowedRoles.Count < 1) 
                            return await Task.FromResult(PreconditionResult.FromSuccess());

                        var tempOutput = new StringBuilder();

                        foreach (var allowedRole in data.AllowedRoles) {
                            var role = context.Guild.GetRole(allowedRole.RoleId);

                            tempOutput.AppendLine($"| {role.Name} |");

                            if (user.RoleIds.Contains(allowedRole.RoleId))
                                return await Task.FromResult(PreconditionResult.FromSuccess());
                        }

                        output.AppendLine("This command is locked to specific role(s). You must have the following role(s)...")
                            .AppendLine()
                            .AppendLine(tempOutput.ToString());
                        break;
                }
            }

            return await Task.FromResult(PreconditionResult.FromError(output.ToString()));
        }
    }
}